### ISSUE TITLE
macOS - Cannot change app dock icon dynamically

### ISSUE THREAD

**Issue by suptejas** [tauri-apps/tauri#2985]
### **Describe the bug**
`window.set_icon()` always throws errors on macOS (with `icns`) and does not set the dock icon. This works fine with Windows (`ico`) and Linux (`png`) in which both the window icon and the dock icon is changed.

As far as I know, this can be done using the dock api on electron with something like:
`dock.setIcon(icon)`

### **To Reproduce**
1. Create a new Tauri app on the latest version of macOS
2. Go to `src-tauri/src/main.rs`
3. Add a `setup(|app|)` function to the app build stage.
4. `let main_window = app.get_window("main").unwrap()`
5. 
```rs
main_window
  .set_icon(Icon::File(
    PathBuf::from_str("<path to icns file>").unwrap(),
  ))
  .unwrap();
```
6. See error:
![image](https://user-images.githubusercontent.com/63039748/143926088-a9c73114-e17e-44aa-9697-dd1c26aefe14.png)
```rs
thread 'main' panicked at 'could not determine icon extension', /Users/xtremedevx/.cargo/registry/src/github.com-1ecc6299db9ec823/tauri-runtime-wry-0.2.1/src/lib.rs:467:8
```

### **Expected behavior**
Since macOS does not have a window icon, there should be a way to change the icon of the app on the dock.

`app.set_dock_icon()` or another method could be used to set the icon of the app on the dock.

### **Platform and Versions (required):**

```
Operating System - Mac OS, version 12.0.1 X64

Node.js environment
  Node.js - 16.13.0
  @tauri-apps/cli - 1.0.0-beta.10
  @tauri-apps/api - Not installed

Global packages
  npm - 8.1.0
  yarn - 1.22.17

Rust environment
  rustc - 1.56.1
  cargo - 1.56.0

App directory structure
/dist
/node_modules
/src-tauri
/.git
/src

App
  tauri.rs - 1.0.0-beta.8
  build-type - bundle
  CSP - default-src blob: data: filesystem: ws: wss: http: https: tauri: 'unsafe-eval' 'unsafe-inline' 'self' img-src: 'self'
  distDir - ../dist
  devPath - ../src
```

**Comment by probablykasper** [comment]
Related: #2358

The `window.set_icon` method probably shouldn't exist on macOS.

Similar to what's been suggested for handling menu bars, it could work like this:
- `App` and `AppHandle` provides a `set_app_icon()` method.
- Linux/Windows only: `Window` provides a `set_icon()` method which overrides the app icon, or resets it to the app icon if `None` is passed.

(Disclaimer: I'm not 100% sure how Linux/Windows window icons work)

**Comment by lucasfernog** [comment]
For future reference, this is the code to switch the app dock icon:

```rust
#[macro_use]
extern crate objc;

fn change_icon() {
  unsafe {
    use cocoa::{
      appkit::NSImage,
      base::{id, nil},
      foundation::NSData,
    };
    let app: id = msg_send![class!(NSApplication), sharedApplication];
    let icon = vec![1, 2, 3]; // icon bytes here
    let data = NSData::dataWithBytes_length_(
      nil,
      icon.as_ptr() as *const std::os::raw::c_void,
      icon.len() as u64,
    );
	let app_icon = NSImage::initWithData_(NSImage::alloc(nil), data);
    let _: () = msg_send![app, setApplicationIconImage: app_icon];
}
```

**Comment by amrbashir** [comment]
closing this as it will be solved by #2358
